qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py

```diff
@@ -0,0 +1,756 @@
+                # Add events and qubits from doomed block to survivor.
+                if survivor.active_qubits.intersection(doomed_delay_group.active_qubits):
+                    raise RuntimeError("More than one open delay on a qubit?")
+
+                survivor.events.extend(doomed_delay_group.events)
+                survivor.active_qubits.update(doomed_delay_group.active_qubits)
+
+                open_delay_blocks.remove(doomed_delay_group)
+                survivor.events.sort(key=DelayEvent.sort_key)  # Maintain sorted event order
+
+        for delay_event in sorted_delay_events:
+            # This could be avoided by keeping a map of device qubit to open
+            # block and only considering neighbors of current event.
+            # use coupling_map.graph.neighbors_undirected once Qiskit/rustworkx#1254 is in a release
+            adjacent_open_delay_blocks = [
+                open_delay
+                for open_delay in open_delay_blocks
+                if any(
+                    self._coupling_map.distance(
+                        qubit_map[delay_event.op_node.qargs[0]], qubit_map[open_delay_qubit]
+                    )
+                    <= 1
+                    for open_delay_qubit in open_delay.active_qubits
+                )
+            ]
+
+            if delay_event.type == EventType.BEGIN:
+                # If crossing a begin edge, check if there are any open delays that are adjacent.
+                # If so, add current event to that group.
+
+                if len(adjacent_open_delay_blocks) == 0:
+                    # Make a new delay
+                    _open_delay_block(delay_event)
+                else:
+                    # Make a new block and combine that with adjacent open blocks
+                    new_block = _open_delay_block(delay_event)
+                    _combine_delay_blocks(adjacent_open_delay_blocks + [new_block])
+
+            if delay_event.type == EventType.END:
+                # If crossing a end edge, remove this qubit from the actively delaying qubits"
+                if len(adjacent_open_delay_blocks) != 1:
+                    raise RuntimeError("Closing edge w/o an open delay?")
+
+                _update_delay_block(adjacent_open_delay_blocks[0], delay_event)
+
+        # log the open delays and the number of closed delays
+        logger.debug("-- post collect")
+        logger.debug("open delays: %s", open_delay_blocks)
+        logger.debug("len(closed delays): %s", len(closed_delay_blocks))
+
+        # validate the results, there should be no open delays and all active qubits
+        # should be accounted for
+        if len(open_delay_blocks) > 0:
+            raise RuntimeError("Failed to close all open delays.")
+
+        for closed_delay in closed_delay_blocks:
+            if len(closed_delay.active_qubits) > 0:
+                raise RuntimeError("Failed remove active qubits on closed delays.")
+
+            if validate:
```

Comment:
There are several validation checks across the code. I suggest enabling and using them in `test_context_aware_dd.py`